pkg/runtime/runtime.go

```diff
@@ -17,13 +17,30 @@ limitations under the License.
 package runtime
 
 import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"iter"
 	"maps"
+	"strings"
 
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	trainer "github.com/kubeflow/trainer/pkg/apis/trainer/v1alpha1"
+	"github.com/kubeflow/trainer/pkg/apply"
+	"github.com/kubeflow/trainer/pkg/constants"
 	corev1 "k8s.io/api/core/v1"
 	corev1ac "k8s.io/client-go/applyconfigurations/core/v1"
 	resourcehelpers "k8s.io/component-helpers/resource"
+)
 
-	trainer "github.com/kubeflow/trainer/pkg/apis/trainer/v1alpha1"
+var (
+	errorTemplateSpecPathNotFound = errors.New("template spec path not found")
+
+	defaultPodSetsSyncer = func(*Info) {}
```

Comment:
How this sync is used ?

Reply:
Here is one of sync flow:

1. Register callback function: https://github.com/kubeflow/trainer/blob/b938a3971ea442942993f65ef400f70af5c9e62b/pkg/runtime/core/trainingruntime.go#L125
2. Obtainer syncer: https://github.com/kubeflow/trainer/blob/b938a3971ea442942993f65ef400f70af5c9e62b/pkg/runtime/runtime.go#L224
3. Sync PodSpec to TemplateSpecApplyConfiguration: https://github.com/kubeflow/trainer/blob/b938a3971ea442942993f65ef400f70af5c9e62b/pkg/runtime/framework/plugins/mpi/mpi.go#L199